File: apps/api-rs/src/document/pptx.rs
use super::shared::*;
use super::*;
pub(super) fn parse_pptx(
    doc_id: Uuid,
    parse_job_id: Uuid,
    title: &str,
    bytes: &[u8],
) -> Result<ParsedDocument> {
    let mut archive = open_zip(bytes)?;
    let (slide_names, used_relationship_order) = pptx_slide_names(&mut archive)?;

    let mut blocks = Vec::new();
    let mut tables = Vec::new();
    let mut anchors = Vec::new();
    let mut warnings = Vec::new();
    if !used_relationship_order {
        warnings.push("pptx_slide_order_fallback".to_string());
    }

    for (slide_idx, name) in slide_names.iter().enumerate() {
        let xml = read_zip_text(&mut archive, name)?;
        validate_xml_nesting(&xml)?;
        let doc = roxmltree::Document::parse(&xml).context("invalid_pptx_slide_xml")?;
        let mut slide_heading: Vec<String> = Vec::new();
        let mut has_text = false;
        let sp_tree = doc
            .descendants()
            .find(|node| node.is_element() && node.tag_name().name() == "spTree");
        let shape_nodes: Vec<_> = sp_tree
            .map(|tree| tree.children().filter(|node| node.is_element()).collect())
            .unwrap_or_default();

        for (shape_idx, shape) in shape_nodes.into_iter().enumerate() {
            if let Some(tbl) = shape
                .descendants()
                .find(|node| node.is_element() && node.tag_name().name() == "tbl")
            {
                let block_id = Uuid::new_v4();
                let table = parse_xml_table(
                    tbl,
                    doc_id,
                    parse_job_id,
                    block_id,
                    tables.len() as i32,
                    slide_heading.clone(),
                    None,
                    None,
                    Some((slide_idx + 1) as i32),
                    "pptx",
                );
                blocks.push(ParsedBlock {
                    block_id,
                    block_index: blocks.len() as i32,
                    block_type: "table".to_string(),
                    text: table.markdown.clone(),
                    heading_level: None,
                    heading_path: table.heading_path.clone(),
                    page_start: None,
                    page_end: None,
                    slide_index: Some((slide_idx + 1) as i32),
                    table_id: Some(table.table_id),
                    bbox: None,
                    anchor_ids: vec![],
                    source_ref: json!({"format": "pptx", "slide": slide_idx + 1, "shape": shape_idx + 1, "node": "a:tbl"}),
                    metadata: json!({}),
                });
                tables.push(table);
                continue;
            }

            let is_title_shape = shape.descendants().any(|node| {
                node.is_element()
                    && node.tag_name().name() == "ph"
                    && attr(node, "type").is_some_and(|kind| matches!(kind, "title" | "ctrTitle"))
            });
            for p in shape
                .descendants()
                .filter(|node| node.is_element() && node.tag_name().name() == "p")
            {
                let text = collect_text(p).trim().to_string();
                if text.is_empty() {
                    continue;
                }
                let is_bullet = p
                    .descendants()
                    .any(|node| matches!(node.tag_name().name(), "buChar" | "buAutoNum"));
                let is_heading = is_title_shape || !has_text;
                let block_type = if is_heading {
                    "heading"
                } else if is_bullet {
                    "list_item"
                } else {
                    "paragraph"
                };
                if is_heading {
                    slide_heading = vec![text.clone()];
                }
                has_text = true;
                blocks.push(ParsedBlock {
                    block_id: Uuid::new_v4(),
                    block_index: blocks.len() as i32,
                    block_type: block_type.to_string(),
                    text,
                    heading_level: is_heading.then_some(1),
                    heading_path: if is_heading { vec![] } else { slide_heading.clone() },
                    page_start: None,
                    page_end: None,
                    slide_index: Some((slide_idx + 1) as i32),
                    table_id: None,
                    bbox: None,
                    anchor_ids: vec![],
                    source_ref: json!({"format": "pptx", "slide": slide_idx + 1, "shape": shape_idx + 1, "node": "a:p"}),
                    metadata: json!({"placeholder_title": is_title_shape}),
                });
            }
        }

        if !has_text
            && !blocks
                .iter()
                .any(|block| block.slide_index == Some((slide_idx + 1) as i32))
        {
            warnings.push(format!("slide_{}_empty", slide_idx + 1));
        }
    }
    attach_table_cell_anchors(
        doc_id,
        parse_job_id,
        "pptx",
        &mut blocks,
        &tables,
        &mut anchors,
    );

    let mut parsed = finalize_parsed(
        doc_id,
        parse_job_id,
        "pptx",
        title,
        Some(slide_names.len() as i32),
        blocks,
        tables,
        anchors,
    );
    parsed.warnings.extend(warnings);
    Ok(parsed)
}

fn pptx_slide_names(archive: &mut ZipArchive<Cursor<&[u8]>>) -> Result<(Vec<String>, bool)> {
    if let (Ok(presentation_xml), Ok(rels_xml)) = (
        read_zip_text(archive, "ppt/presentation.xml"),
        read_zip_text(archive, "ppt/_rels/presentation.xml.rels"),
    ) {
        validate_xml_nesting(&presentation_xml)?;
        validate_xml_nesting(&rels_xml)?;
        let presentation = roxmltree::Document::parse(&presentation_xml)
            .context("invalid_pptx_presentation_xml")?;
        let relationships =
            roxmltree::Document::parse(&rels_xml).context("invalid_pptx_relationships_xml")?;
        let targets = relationships
            .descendants()
            .filter(|node| node.is_element() && node.tag_name().name() == "Relationship")
            .filter_map(|node| Some((attr(node, "Id")?, attr(node, "Target")?)))
            .collect::<std::collections::BTreeMap<_, _>>();
        let ordered = presentation
            .descendants()
            .filter(|node| node.is_element() && node.tag_name().name() == "sldId")
            .filter_map(|node| {
                node.attributes()
                    .find(|attribute| {
                        attribute.name() == "id" && attribute.value().starts_with("rId")
                    })
                    .map(|attribute| attribute.value())
            })
            .filter_map(|relationship_id| targets.get(relationship_id))
            .map(|target| normalize_ppt_target(target))
            .collect::<Vec<_>>();
        if !ordered.is_empty() {
            return Ok((ordered, true));
        }
    }

    let mut slide_names: Vec<String> = (0..archive.len())
        .filter_map(|i| archive.by_index(i).ok().map(|f| f.name().to_string()))
        .filter(|name| {
            name.starts_with("ppt/slides/slide")
                && name.ends_with(".xml")
                && name["ppt/slides/slide".len()..]
                    .trim_end_matches(".xml")
                    .chars()
                    .all(|c| c.is_ascii_digit())
        })
        .collect();
    slide_names.sort_by_key(|name| {
        name.trim_start_matches("ppt/slides/slide")
            .trim_end_matches(".xml")
            .parse::<i32>()
            .unwrap_or(0)
    });

    Ok((slide_names, false))
}

fn normalize_ppt_target(target: &str) -> String {
    let target = target.trim_start_matches('/');
    if target.starts_with("ppt/") {
        target.to_string()
    } else {
        format!("ppt/{}", target.trim_start_matches("../"))
    }
}

File: apps/api-bun/src/document/pptx.ts
// 移植自 apps/api-rs/src/document/pptx.rs —— 行为对齐 Rust 原版，错误信息保持一致

// PPT：优先按 ppt/presentation.xml + rels 的 sldId 顺序取幻灯片，退化时按文件名排序扫描

import type JSZip from 'jszip';
import { newUuid } from '../infra/uuid.ts';
import type { SourceAnchor } from '../models/source_anchor.ts';
import {
  attachTableCellAnchors,
  collectText,
  finalizeParsed,
  openZip,
  parseXmlTable,
  readZipText,
} from './shared.ts';
import {
  attr,
  childElements,
  descendants,
  parseXml,
  validateXmlNesting,
  type XmlElement,
} from './xml.ts';
import { trimRust } from './text_utils.ts';
import type { ParsedBlock, ParsedDocument, ParsedTable } from './types.ts';

export async function parsePptx(
  docId: string,
  parseJobId: string,
  title: string,
  bytes: Uint8Array,
): Promise<ParsedDocument> {
  const archive = await openZip(bytes);
  const [slideNames, usedRelationshipOrder] = await pptxSlideNames(archive);

  const blocks: ParsedBlock[] = [];
  const tables: ParsedTable[] = [];
  const anchors: SourceAnchor[] = [];
  const warnings: string[] = [];
  if (!usedRelationshipOrder) {
    warnings.push('pptx_slide_order_fallback');
  }

  for (const [slideIdx, name] of slideNames.entries()) {
    const slideIndex = slideIdx + 1;
    const xml = await readZipText(archive, name);
    validateXmlNesting(xml);
    const doc = parseXml(xml);
    let slideHeading: string[] = [];
    let hasText = false;
    const spTree = descendants(doc)
      .find((ref) => ref.kind === 'element' && ref.element.name === 'spTree');
    const shapeNodes = spTree !== undefined && spTree.kind === 'element'
      ? childElements(spTree.element)
      : [];

    for (const [shapeIdx, shape] of shapeNodes.entries()) {
      const shapeIndex = shapeIdx + 1;
      const tbl = descendants(shape)
        .find((ref) => ref.kind === 'element' && ref.element.name === 'tbl');
      if (tbl !== undefined && tbl.kind === 'element') {
        const blockId = newUuid();
        const table = parseXmlTable(
          tbl.element,
          docId,
          parseJobId,
          blockId,
          tables.length,
          [...slideHeading],
          null,
          null,
          slideIndex,
          'pptx',
        );
        blocks.push({
          block_id: blockId,
          block_index: blocks.length,
          block_type: 'table',
          text: table.markdown,
          heading_level: null,
          heading_path: [...table.heading_path],
          page_start: null,
          page_end: null,
          slide_index: slideIndex,
          table_id: table.table_id,
          bbox: null,
          anchor_ids: [],
          source_ref: { format: 'pptx', slide: slideIndex, shape: shapeIndex, node: 'a:tbl' },
          metadata: {},
        });
        tables.push(table);
        continue;
      }

      const isTitleShape = descendants(shape).some((ref) => {
        if (ref.kind !== 'element' || ref.element.name !== 'ph') return false;
        const kind = attr(ref.element, 'type');
        return kind === 'title' || kind === 'ctrTitle';
      });
      for (const ref of descendants(shape)) {
        if (ref.kind !== 'element' || ref.element.name !== 'p') continue;
        const text = trimRust(collectText(ref.element));
        if (text.length === 0) continue;
        const isBullet = descendants(ref.element).some(
          (inner) => inner.kind === 'element' && (inner.element.name === 'buChar' || inner.element.name === 'buAutoNum'),
        );
        const isHeading = isTitleShape || !hasText;
        const blockType = isHeading ? 'heading' : isBullet ? 'list_item' : 'paragraph';
        if (isHeading) {
          slideHeading = [text];
        }
        hasText = true;
        blocks.push({
          block_id: newUuid(),
          block_index: blocks.length,
          block_type: blockType,
          text,
          heading_level: isHeading ? 1 : null,
          heading_path: isHeading ? [] : [...slideHeading],
          page_start: null,
          page_end: null,
          slide_index: slideIndex,
          table_id: null,
          bbox: null,
          anchor_ids: [],
          source_ref: { format: 'pptx', slide: slideIndex, shape: shapeIndex, node: 'a:p' },
          metadata: { placeholder_title: isTitleShape },
        });
      }
    }

    if (!hasText && !blocks.some((block) => block.slide_index === slideIndex)) {
      warnings.push(`slide_${slideIndex}_empty`);
    }
  }

  attachTableCellAnchors(docId, parseJobId, 'pptx', blocks, tables, anchors);

  const parsed = finalizeParsed(docId, parseJobId, 'pptx', title, slideNames.length, blocks, tables, anchors);
  parsed.warnings.push(...warnings);
  return parsed;
}

async function pptxSlideNames(archive: JSZip): Promise<[string[], boolean]> {
  const presentationXml = await readZipTextOptional(archive, 'ppt/presentation.xml');
  const relsXml = await readZipTextOptional(archive, 'ppt/_rels/presentation.xml.rels');
  if (presentationXml !== null && relsXml !== null) {
    validateXmlNesting(presentationXml);
    validateXmlNesting(relsXml);
    const presentation = parseXmlOrThrow(presentationXml, 'invalid_pptx_presentation_xml');
    const relationships = parseXmlOrThrow(relsXml, 'invalid_pptx_relationships_xml');
    const targets = new Map<string, string>();
    for (const ref of descendants(relationships)) {
      if (ref.kind !== 'element' || ref.element.name !== 'Relationship') continue;
      const id = attr(ref.element, 'Id');
      const target = attr(ref.element, 'Target');
      if (id === undefined || target === undefined) continue;
      targets.set(id, target);
    }
    const ordered: string[] = [];
    for (const ref of descendants(presentation)) {
      if (ref.kind !== 'element' || ref.element.name !== 'sldId') continue;
      // attrs 是扁平本地名映射：<p:sldId id="256" r:id="rId2"/> 取到的是 r:id
      const relationshipId = Object.entries(ref.element.attrs)
        .find(([name, value]) => name === 'id' && value.startsWith('rId'))?.[1];
      if (relationshipId === undefined) continue;
      const target = targets.get(relationshipId);
      if (target === undefined) continue;
      ordered.push(normalizePptTarget(target));
    }
    if (ordered.length > 0) {
      return [ordered, true];
    }
  }

  const slideNames: string[] = [];
  for (const file of Object.values(archive.files)) {
    const name = file.name;
    if (!name.startsWith('ppt/slides/slide') || !name.endsWith('.xml')) continue;
    const rest = trimEndMatches(name.slice('ppt/slides/slide'.length), '.xml');
    if (!/^[0-9]*$/.test(rest)) continue;
    slideNames.push(name);
  }
  slideNames.sort((a, b) => slideOrderKey(a) - slideOrderKey(b));

  return [slideNames, false];
}

async function readZipTextOptional(archive: JSZip, name: string): Promise<string | null> {
  try {
    return await readZipText(archive, name);
  } catch {
    return null;
  }
}

function parseXmlOrThrow(xml: string, context: string): XmlElement {
  try {
    return parseXml(xml);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new Error(`${context}:${message}`);
  }
}

function trimEndMatches(value: string, pattern: string): string {
  let out = value;
  while (out.endsWith(pattern)) out = out.slice(0, -pattern.length);
  return out;
}

function trimStartMatches(value: string, pattern: string): string {
  let out = value;
  while (out.startsWith(pattern)) out = out.slice(pattern.length);
  return out;
}

function slideOrderKey(name: string): number {
  const raw = trimEndMatches(trimStartMatches(name, 'ppt/slides/slide'), '.xml');
  if (!/^[+-]?\d+$/.test(raw)) return 0;
  const parsed = Number.parseInt(raw, 10);
  return parsed < -2147483648 || parsed > 2147483647 ? 0 : parsed;
}

function normalizePptTarget(target: string): string {
  const trimmed = trimStartMatches(target, '/');
  if (trimmed.startsWith('ppt/')) return trimmed;
  return 'ppt/' + trimStartMatches(trimmed, '../');
}
